authenticator-rs: init at 0.7.6

philiptaron opened the pull request:
## Description of changes

[authenticator-rs](https://github.com/grumlimited/authenticator-rs) is a TOTP-MFA application written in Rust and GTK3.

It is currently packaged as a .deb, an .rpm, and is in the [AUR](https://aur.archlinux.org/packages/authenticator-rs-bin).

### Main view

![authenticator-rs](https://github.com/grumlimited/authenticator-rs/blob/master/data/screenshots/screenshot1.png?raw=true "Main view")
![authenticator-rs](https://github.com/grumlimited/authenticator-rs/blob/master/data/screenshots/screenshot2.png?raw=true "Main view")

### Group editing

![authenticator-rs](https://github.com/grumlimited/authenticator-rs/blob/master/data/screenshots/screenshot3.png?raw=true "Group editing")

### Account editing

![authenticator-rs](https://github.com/grumlimited/authenticator-rs/blob/master/data/screenshots/screenshot4.png?raw=true "Account editing")

### Importing and exporting

Using `yaml` format

    ---
    - name: group name
      url: bbc.co.uk
      entries:
        - label: Account 1
          secret: secret code

![authenticator-rs](https://github.com/grumlimited/authenticator-rs/blob/master/data/screenshots/screenshot5.png?raw=true "Importing")

## Things done

Packaging this application was a little more challenging than I expected.

* The version numbers in the repository didn't match the version numbers in the metadata. The metadata was 0.1 and 0.0.1; the repository is 0.7.6.

* Two tests make use of the network.

* The install process uses a hand-written Makefile that baked in assumptions about `/usr`. It also re-compiled the source.

I also decided to strip "RS" from the `.desktop` name, as it didn't add anything to the functionality of the application.

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

philiptaron commented:
@matthiasbeyer, as a new committer and a Rust aficionado, perhaps you'd like to review and merge if it passes your bar?

matthiasbeyer commented on pkgs/by-name/au/authenticator-rs/fix-install.patch:
```diff
@@ -0,0 +1,133 @@
+ 	# Force icon cache refresh
+ 	touch $(sharedir)/icons/hicolor
+ 
+-	# Install application meta-data
++	# Install application metadata
```
This patch fragment is not necessary.

matthiasbeyer commented on pkgs/by-name/au/authenticator-rs/fix-install.patch:
```diff
@@ -0,0 +1,133 @@
+-	meson builddir --prefix=/usr
++	meson setup --prefix=$(PREFIX) builddir
+ 	DESTDIR=../build-aux/i18n meson install -C builddir
+-	cp -fr build-aux/i18n/usr $(DESTDIR)
```
Why was this removed?

philiptaron replied:
This is removed in https://github.com/grumlimited/authenticator-rs/pull/138 in order to have the `meson install` line do the copy itself. We need to do that since the current working directory doesn't have `build-aux` when invoked by Nix.

matthiasbeyer commented on pkgs/by-name/au/authenticator-rs/fix-install.patch:
```diff
@@ -0,0 +1,133 @@
+ 
+ #### Arch Linux
+ 
+-Authenticator RS is also available from [AUR](https://aur.archlinux.org/packages/authenticator-rs-bin/). Or from the release page:
++Authenticator is also available from [AUR](https://aur.archlinux.org/packages/authenticator-rs-bin/). Or from the release page:
```
This diff is also not necessary.

matthiasbeyer commented on pkgs/by-name/au/authenticator-rs/fix-install.patch:
```diff
@@ -0,0 +1,133 @@
+                 <property name="valign">center</property>
+                 <property name="hexpand">True</property>
+                 <property name="vexpand">True</property>
+-                <property name="label">Authenticator RS</property>
++                <property name="label">Authenticator</property>
```
IMO this is not necessary. Also it renames the application from what the authors wanted, I would object that.

matthiasbeyer commented on pkgs/by-name/au/authenticator-rs/fix-install.patch:
```diff
@@ -0,0 +1,133 @@
+             <property name="valign">center</property>
+             <property name="hexpand">True</property>
+             <property name="vexpand">True</property>
+-            <property name="label">Authenticator RS</property>
++            <property name="label">Authenticator</property>
```
Same as above.

matthiasbeyer commented on pkgs/by-name/au/authenticator-rs/fix-install.patch:
```diff
@@ -0,0 +1,133 @@
++++ b/data/uk.co.grumlimited.authenticator-rs.desktop
+@@ -1,5 +1,5 @@
+ [Desktop Entry]
+-Name=Authenticator RS
++Name=Authenticator
```
Same as above.

matthiasbeyer commented on pkgs/by-name/au/authenticator-rs/fix-install.patch:
```diff
@@ -0,0 +1,133 @@
+ 
+         gui.set_application(app, connection);
+ 
+-        info!("Authenticator RS initialised");
++        info!("Authenticator initialised");
```
Same as above.

matthiasbeyer reviewed (changes requested):
Hi!

Awesome, thanks for contributing this package!

There are a few things that IMO have to be changed: First of all, the patch file introduces more changes than necessary. Keep the changed stuff from patches as minimal as possible. For example, removing the "RS" prefix - that's renaming the application from what the original authors wanted it to be named. We shouldn't do that.

Also, we might want to see whether we can contribute these changes upstream - what do you think? The changes to the `Makefile` could be contributed upstream I guess... maybe they'll be accepted?

From the nix code itself, this PR looks fine IMO! :tada: Thanks again for taking the time for contributing!

philiptaron committed:
authenticator-rs: init at 0.7.6

Packaging this application was a little more challenging than I expected,
but not too bad. I learned a lot!

* The version numbers in the repository didn't match the version numbers
  in the metadata. The metadata was 0.1 and 0.0.1; the repository is 0.7.6.
  PR upstream: https://github.com/grumlimited/authenticator-rs/pull/139

  For this revision, I use `sed` instead of of pulling in the patch
  upstream, since it needs to re-generate Cargo.lock and I didn't want
  to do that as part of packaging for nixpkgs.

* The install process uses a hand-written Makefile that baked in
  assumptions about `/usr`. It also re-compiled the source.
  PR upstream: https://github.com/grumlimited/authenticator-rs/pull/138

philiptaron commented:
Thank you so much for the review, @matthiasbeyer. I've opened two PRs against upstream for the issues I ran into while packaging, and the Nixpkgs commit now no longer holds a patch.

> There are a few things that IMO have to be changed: First of all, the patch file introduces more changes than necessary. Keep the changed stuff from patches as minimal as possible. For example, removing the "RS" prefix - that's renaming the application from what the original authors wanted it to be named. We shouldn't do that.

Removed all branding changes. That can stay in my personal NixOS configuration.

> Also, we might want to see whether we can contribute these changes upstream - what do you think? The changes to the `Makefile` could be contributed upstream I guess... maybe they'll be accepted?

Hope so!

> From the nix code itself, this PR looks fine IMO! 🎉 Thanks again for taking the time for contributing!

Thanks for the review, and I think it's ready for a re-review.

matthiasbeyer commented:
Can you please share a link to the PRs against upstream? Would be nice to track these.

philiptaron commented:
> Can you please share a link to the PRs against upstream? Would be nice to track these.

Sure thing. They're in the comments, and also here:

1. https://github.com/grumlimited/authenticator-rs/pull/138
2. https://github.com/grumlimited/authenticator-rs/pull/139
